src/backend/distributed/transaction/worker_transaction.c

```diff
@@ -0,0 +1,479 @@
+#include "distributed/worker_transaction.h"
+#include "utils/memutils.h"
+
+
+/* Local functions forward declarations */
+static List * OpenWorkerTransactions(void);
+static TransactionConnection * GetWorkerTransaction(char *nodeName, int32 nodePort);
+static bool IsResponseOK(ExecStatusType resultStatus);
+static void EnableXactCallback(void);
+static void CompleteWorkerTransactions(XactEvent event, void *arg);
+
+
+/* Global worker connection list */
+static List *workerConnectionList = NIL;
+static bool isXactCallbackRegistered = false;
+
+
+/*
+ * GetWorkerTransactions opens connections to all workers and starts
+ * a transaction block that is committed or aborted when the local
+ * transaction commits or aborts. Multiple invocations of
+ * GetWorkerTransactions within the same transaction will return
+ * the same list of connections.
+ */
+List *
+GetWorkerTransactions(void)
+{
+	if (workerConnectionList == NIL)
+	{
+		InitializeDistributedTransaction();
+		EnableXactCallback();
+
+		workerConnectionList = OpenWorkerTransactions();
+	}
+
+	/* ensure that number of workers has not change */
+	Assert(list_length(WorkerNodeList()) == list_length(workerConnectionList));
+
+	return workerConnectionList;
+}
+
+
+/*
+ * SendCommandToWorker sends a command to a particular worker as part of the
+ * 2PC.
+ */
+void
+SendCommandToWorker(char *nodeName, int32 nodePort, char *command)
+{
+	TransactionConnection *transactionConnection = NULL;
+	PGresult *queryResult = NULL;
+	ExecStatusType resultStatus = PGRES_EMPTY_QUERY;
+
+	transactionConnection = GetWorkerTransaction(nodeName, nodePort);
+	if (transactionConnection == NULL)
+	{
+		ereport(ERROR, (errmsg("worker %s:%d is not part of current transaction",
+							   nodeName, nodePort)));
+	}
+
+	queryResult = PQexec(transactionConnection->connection, command);
+	resultStatus = PQresultStatus(queryResult);
+	if (resultStatus != PGRES_COMMAND_OK && resultStatus != PGRES_TUPLES_OK)
+	{
+		ReraiseRemoteError(transactionConnection->connection, queryResult);
+	}
+
+	PQclear(queryResult);
+}
+
+
+/*
+ * SendCommandToWorkersInParallel sends a command to all workers in
+ * parallel. Commands are committed on the workers when the local
+ * transaction commits. The connection are made as the extension
+ * owner to ensure write access to the Citus metadata tables.
+ */
+void
+SendCommandToWorkersInParallel(char *command)
+{
+	SendCommandToWorkersInParallelParams(command, 0, NULL, NULL);
+}
+
+
+/*
+ * SendCommandToWorkersInParallelParams sends a command to all workers in parallel.
+ * Commands are committed on the workers when the local transaction commits. The
+ * connection are made as the extension owner to ensure write access to the Citus
+ * metadata tables. Parameters can be specified as for PQexecParams, except that
+ * paramLengths, paramFormats and resultFormat are hard-coded to NULL, NULL and 0
+ * respectively.
+ */
+void
+SendCommandToWorkersInParallelParams(char *command, int parameterCount,
```

Comment:
@marcocitus SendCommandToWorkersInParallel function can be beneficial for other use cases than metadata syncing. You are correct for the metadata syncing commands but for non-metadata commands, we may want to send an arbitrary query to the worker nodes.

I suggest adding a parameter to the function indication whether the command is a metadata command, or arbitrary command. What do you think?